### PR TITLE
Fix num_alert_types

### PR DESCRIPTION
**Pull request by aldenml**
Fix num_alert_types.

### CODE REVIEW

File: include/libtorrent/alert_types.hpp [hunk 1]
```diff
@@ -2496,7 +2537,7 @@ namespace libtorrent
 #undef TORRENT_DEFINE_ALERT_PRIO
 #undef TORRENT_CLONE
 
-	enum { num_alert_types = 89 };
+	enum { num_alert_types = 90 };
```
Comment (arvidn) [inline include/libtorrent/alert_types.hpp]: I get this fix.. what about the other changes?

Comment (aldenml) [inline include/libtorrent/alert_types.hpp]: The other changes are beacouse `num_alert_types` is not dynamic. It seems to me a little odd not having the right amount of alert types even if `TORRENT_DISABLE_LOGGING` is defined. Specifically for picker_log_alert there was no constructor if `TORRENT_DISABLE_LOGGING` and `peer_alert` has no default constructor.

**Comment by aldenml** [comment]
rebasing.

File: include/libtorrent/alert_types.hpp (hunk 1 above)
Comment (arvidn) [inline include/libtorrent/alert_types.hpp]: ah. the intention of this enum is really "max_alert_index" + 1. It makes it possible to use alert types as indices into a dense array. If some alerts will never actually be posted is not really relevant for that purpose.
are you using it for anything where it's important that it indicates the actual number of alerts that exists?
do you think it should be called something else?

Comment (aldenml) [inline include/libtorrent/alert_types.hpp]: Actually, I'm compiling with logging enabled. Before give up with this PR in the current form, I will explain a problem I'm facing (see in the PR's comment).

**Comment by aldenml** [comment]
The main motivation of this PR is that I'm dealing with a way to efficiently cast the alerts in a swig/java world. Having the types in a consecutive sequence allow the java compiler apply heavy optimizations in a dense switch/case statement. This optimization can vary with a function of how many cases and how many holes are in the cases.

**Comment by aldenml** [comment]
I will resubmit this PR with a reduced set of changes.

**Pull request merged**